lib/streamlit/server/server.py

```diff
@@ -396,8 +406,38 @@ def _set_state(self, new_state):
         self._state = new_state
 
     @property
-    def is_ready_for_browser_connection(self):
-        return self._state not in (State.INITIAL, State.STOPPING, State.STOPPED)
+    def is_ready_for_browser_connection(self) -> Tuple[bool, str]:
+        if self._state not in (State.INITIAL, State.STOPPING, State.STOPPED):
+            return True, "ok"
+
+        return False, "unavailable"
+
+    @property
+    def is_script_loading(self) -> Tuple[bool, str]:
+        try:
+            with source_util.open_python_file(self.script_path) as f:
+                filebody = f.read()
+
+            if config.get_option("runner.magicEnabled"):
+                filebody = magic.add_magic(filebody, self.script_path)
+
+            code = compile(
+                filebody,
+                self.script_path,
+                mode="exec",
+                flags=0,
+                dont_inherit=1,
+                optimize=-1,
+            )
+
+            module = types.ModuleType("__main__")
+            sys.modules["__main__"] = module
+            module.__dict__["__file__"] = self.script_path
+            exec(code, module.__dict__)
```

Comment:
I'm a bit concerned about what might happen when actually trying to run the script here.

Because running a streamlit script requires a decent amount of machinery to be in place, we use an internal global variable `_is_running_with_streamlit` to ensure that a script is essentially a big no-op when a user does something like running it directly (that is, runs something like `python3 my_script.py`).

This flag gets set to true from within `cli.py`. What could be strange about running a script from here is that the flag may be set when the script is run from here, but in this context, I don't believe most of the machinery required to actually run many of the `st.*` commands will be available.

I'm not exactly sure what'll happen when running the script in `exec`, so it would be useful to check the value of `st._is_running_with_streamlit` from within the script when it's run via this endpoint. If the flag is set to True, then it's possible that we'll see some weird/undefined behavior here.

Reply:
I suspect the more-correct thing to do here is to actually kick of a `ScriptRunner` execution of the script, rather than just `exec`'ing it "raw". That way the script is actually run "within Streamlit" (and if the definition of "within Streamlit" changes in ScriptRunner, it won't need to also change here).

The downside is that using ScriptRunner is a bigger pain than what you've already done (because it requires some inputs, and also it functions asynchronously on another thread that it creates). But we could definitely help with that integration!